addons/sale_project/models/sale_order.py

```diff
@@ -88,45 +95,66 @@ def _action_confirm(self):
-                'project.act_project_project_2_project_task_all')
-            action['domain'] = [('id', 'in', self.tasks_ids.ids)]
-            if action.get('context'):
-                eval_context = self.env['ir.actions.actions']._get_eval_context()
-                eval_context.update({'active_id': task_projects.id})
-                action_context = safe_eval(action['context'], eval_context)
-                action_context.update(eval_context)
-                action['context'] = action_context
-        else:
-            action = self.env["ir.actions.actions"]._for_xml_id("project.action_view_task")
-            action['context'] = {}  # erase default context to avoid default filter
-            if len(self.tasks_ids) > 1:  # cross project kanban task
-                action['views'] = [[False, 'kanban'], [list_view_id, 'tree'], [form_view_id, 'form'], [False, 'graph'], [False, 'calendar'], [False, 'pivot']]
-            elif len(self.tasks_ids) == 1:  # single task -> form view
-                action['views'] = [(form_view_id, 'form')]
-                action['res_id'] = self.tasks_ids.id
+        default_line = next(filter(lambda sol: sol.is_service and sol.product_id.service_policy != 'delivered_manual', self.order_line))
+
+        action = self.env["ir.actions.actions"]._for_xml_id("project.action_view_task")
+        action['context'] = {}  # erase default context to avoid default filter
+        if self.tasks_count > 1:  # cross project kanban task
+            action['views'] = [[kanban_view_id, 'kanban'], [list_view_id, 'tree'], [form_view_id, 'form'], [False, 'graph'], [False, 'calendar'], [False, 'pivot']]
+        else:  # 1 or 0 tasks -> form view
+            action['views'] = [(form_view_id, 'form')]
+            action['res_id'] = self.tasks_ids.id
         # filter on the task of the current SO
         action.setdefault('context', {})
-        action['context'].update({'search_default_sale_order_id': self.id})
+        # set default project
+        default_project_id = default_line.project_id.id or self.project_id.id or (len(self.project_ids) > 0 and self.project_ids[0].id)
+
+        action['context'].update({
+            'search_default_sale_order_id': self.id,
+            'default_sale_order_id': self.id,
+            'default_sale_line_id': default_line.id,
+            'default_partner_id': self.partner_id.id,
+            'default_project_id': default_project_id,
+            'default_user_ids': [self.env.uid],
+        })
         return action
 
     def action_view_project_ids(self):
         self.ensure_one()
+        if not self.order_line:
+            return {'type': 'ir.actions.act_window_close'}
+
+        sorted_line = self.order_line.sorted(lambda sol: sol.is_service and sol.product_id.service_policy != 'delivered_manual' and sol.sequence)
+        default_line = next(filter(lambda sol: sol.is_service and sol.product_id.service_policy != 'delivered_manual' and sol.sequence, sorted_line))
```

Comment:
To use the `Model.sorted` method, you need to give it a method that will fetch the key to sort the records by (see [the documentation](https://www.odoo.com/documentation/16.0/developer/reference/backend/orm.html?highlight=sorted#odoo.models.Model.sorted)):

```suggestion
        sorted_line = self.order_line.sorted(key=lambda sol: sol.sequence)
        default_line = next(filter(lambda sol: sol.is_service and sol.product_id.service_policy != 'delivered_manual', sorted_line))
```

Also FYI, since the filter was a bit confusing to you, you can replace it with a generator expression, like this:
```suggestion
        sorted_line = self.order_line.sorted(key=lambda sol: sol.sequence)
        default_line = next(sol for sol in sorted_line if sol.is_service and sol.product_id.service_policy != 'delivered_manual')
```
It's lazy and a slightly more compact, but it might be less immediately apparent that the point of the line is to filter, so up to you :slightly_smiling_face: